### ISSUE TITLE
[RFC] Data Format

### ISSUE THREAD

**Issue by VoVAllen** [dmlc/dgl#758]
## 🚀 Feature
Save/Load DGLGraph to/from files
https://github.com/dmlc/dgl/pull/728

## Motivation
Since pickling a DGLGraph is slow, we would like to have our own format to store the DGLGraphs with features. 

## Implementations
Storage format
```
The storage structure is
{
  // MetaData Section
  uint64_t kDGLSerializeMagic
  uint64_t kVersion
  uint64_t GraphType
  ** Reserved Area till 4kB **
  dgl_id_t num_graphs
  vector<dgl_id_t> graph_indices (start address of each graph)
  vector<dgl_id_t> nodes_num_list (list of number of nodes for each graph)
  vector<dgl_id_t> edges_num_list (list of number of edges for each graph)
  vector<GraphData> graph_datas;
}
Storage of GraphData is
{
  // Everything uses in csr
  NDArray indptr
  NDArray indices
  NDArray edge_ids
  vector<pair<string, NDArray>> node_tensors;
  vector<pair<string, NDArray>> edge_tensors;
}
```

API
```python
from dgl.graph_serialize import save_graphs, load_graphs
g_list=[...list of DGLGraph]
# Save
save_graphs(filename, g_list)
# Load
g_list=load_graphs(filename)
```

# Todo
Current storage didn't store the graph-level data, such as the graph label. This would be related to the [RFC of graph level data](https://github.com/dmlc/dgl/issues/737). One workaround can be we store the label data separately, since not every graph has labels. However, it's always better to store them together. 

My proposal is:
Label can be a dictionary of tensors, and `assert tensor.shape[0]==number_of_graphs`. The label dict would be stored before graph_data part. And add API `load_metadata` to return the labels, nodes_num_list, and edges_num_list, in case of user want to use them to split the data. 

Feel free to leave any comment.

**Comment by keldLundgaard** [comment]
It might be useful to look into using msgpack for serialize/deserialize. It is portable and can be much faster than pickle.

**Comment by jermainewang** [comment]
My concern is about the `node_tensors` and `edge_tensors` in `GraphData`. Since all the graphs share the same keys and array shapes, these fields are stored repeatedly. Also the `nodes_num_list` and `edges_num_list` can be inferred from `indptr` and `indices` arrays in each graph.

For graph data, we could assume that they will be stacked and stored in one NDArray. We can append a new field `vector<pair<string, NDArray>>` after `graph_data`. You said that "not all the graphs have labels", which confused me. Is there an example? Also, I think it's necessary to have a look at all the TuDortmund datasets.

@keldLundgaard thanks for the pointer. MsgPack looks interesting and the ability to compress small integers are quite useful. My question is how costly is it to compress a large integer array? In our case, a graph is represented by three integer arrays and they are stored in a consecutive memory block. Thus, we can take the raw data pointer and directly write it to disk in one shot. From my understanding, if I want to use the compression technique in MsgPack, I need to iterate over the array and write one integer at a time, which might be slow. Correct me if I'm wrong.

**Comment by VoVAllen** [comment]
> My concern is about the node_tensors and edge_tensors in GraphData. Since all the graphs share the same keys and array shapes, these fields are stored repeatedly.
The major reason is adding an extra string is always negligible comparing the size of tensors. Also this makes the API `save_graphs(graph_list)` easier to understand (without telling user all graphs should have same fields and same feature shape).

> Also the nodes_num_list and edges_num_list can be inferred from indptr and indices arrays in each graph.
I assume user may want feature like selecting graphs into one batch having similar sizes, or do stratified split based on label. Therefore a `loadMetaData` API could return nodes, edge nums, and labels.

> For graph data, we could assume that they will be stacked and stored in one NDArray. We can append a new field vector<pair<string, NDArray>> after graph_data. You said that "not all the graphs have labels", which confused me. Is there an example?
Agree. Just ignore what I said. All graph in one dataset should either all have labels or none has label.

**Comment by jermainewang** [comment]
I don't have further questions, and this RFC looks good to me as long as we could easily extend it in the future. Has this RFC all been implemented in #728 ?
